File: cpp/Decry.cpp
// Decry.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>
#include <iterator>
#include <map>
#include <set>
#include <fstream>
#include <thread>
#include <mutex>
#include "Config.h"



using namespace std;




#include "Utils.h"
#include "SData.h"
aindex_t cipher[TXT_LEN];

std::vector<std::string> split_to_vec(const std::string & s, char delimiter)
{
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(s);
    while (std::getline(tokenStream, token, delimiter))
    {
        tokens.push_back(token);
    }
    return tokens;
}

void split_to_array(const string& s, char delimiter, aindex_t* tokens, size_t len=TXT_LEN) {
    std::string s_token;
    std::istringstream tokenStream(s);
    auto cnt = 0;
    while (std::getline(tokenStream, s_token, delimiter))
    {
        tokens[cnt] = aindex_t(std::stoi(s_token));
        cnt++;
        if (cnt >= len) {
            return;
        }
    }
    return;
}



void read_cipher(aindex_t* cipher_v) {
    string cs;
    getline(cin, cs);
    split_to_array(cs,',',cipher_v);
    // vector<string> csv = split(cs, ',');
    // transform(csv.begin(), csv.end(), std::back_inserter(cipher_v),
    //    [](const std::string& str) { return aindex_t(std::stoi(str)); });
}





void eval(const aindex_t* cipher,const aindex_t* plain,set<aindex_t> *tbl,bool& fault) {
    // seen is a reverse tbl, mapping cipher token to plaintext token
    map<aindex_t,aindex_t> seen{};
    for (auto i = 0; i < TXT_LEN; i++) {
        aindex_t ci_v = cipher[i];
        aindex_t p_v = plain[i];
        auto itr = seen.find(ci_v);
        if (itr==seen.end()) {
            // First time seeing this token
            tbl[p_v].insert(ci_v);
            if (tbl[p_v].size() > rowlen[p_v]) {
                fault = true;
                return;
            }
            seen[ci_v] = p_v;
        }
        else {
            // This token was already seen
            aindex_t p_e = itr->second;
            if (p_e != p_v) {
                // CONTRADICTION!
                fault = true;
                return;
            }
        }
       
    }
}

bool decrypt_type1(const aindex_t* cipher) {
    bool is_type1 = false;
    for (auto idx = 0; idx < PLAINTXT_DICT_SIZE; idx++) {
        bool not_same = false;
        set<aindex_t> tbl[27];
        eval(cipher, plaintxts[idx], tbl, not_same);
        if (!not_same) {
            std::cout << plain_strs[idx] << endl;
            return true;
        }
    }
    return false;
}

typedef struct State {
    int* wstate;
    int* wslen;
    aindex_t* ptxt;
    int* plen;
    int* hits;
    set<aindex_t> *tbl;
    map<aindex_t, aindex_t> *seen;
    State(int* wstate,int* wslen,aindex_t* ptxt,int* plen,int* hits,set<aindex_t> *tbl,map<aindex_t,aindex_t> *seen) : wstate(wstate), wslen(wslen),ptxt(ptxt),plen(plen),hits(hits),tbl(tbl),seen(seen){}
    void pp_seen() {
        cout << "SEEN:\n";
        for (auto x : *seen) {
            cout << x.first << "->" << id2c[x.second] << ", ";
        }
        cout << "\n";
    }
    void pp_tbl() {
        cout << "TBL:\n";
        for (auto i = 0; i < 27; i++) {
            if (tbl[i].size() > 0) {
                cout << id2c[i] << " : ";
                for (aindex_t x : tbl[i]) {
                    cout << x << ", ";
                }
                cout << "\n";
            }
        }
       // cout << "END TBL" << "\n";
    }
    bool consume(aindex_t plc) {
       
        aindex_t c_v = cipher[*plen];
        if (hits[c_v] > 0) {
            aindex_t p_e = (*seen)[c_v];
            if (p_e != plc) {
                // Bad
                return false;
            }
            else {
                hits[c_v] += 1;
                ptxt[*plen] = p_e;
                (*plen) += 1;
                return true;
            }
        }
        else {
            if (tbl[plc].size() + 1 > rowlen[plc]) {
                // Bad
                return false;
            }
            else {
                (*seen)[c_v] = plc;
                tbl[plc].insert(c_v);
                hits[c_v] += 1;
                ptxt[*plen] = plc;
                (*plen) += 1;
                return true;
            }
        }
    }
    void rollback(int num) {
     /*   cout << "Rollback with num: " << num << endl;
        for (auto i = 0; i < ROW_SUM; i++) {
            cout << hits[i] << "|";
        }
        cout << endl;
        cout << "[" << (*plen) << "]Rmv:" ;*/
        for (auto i = 0; i < num; i++) {
            aindex_t c_v = cipher[(*plen) - i - 1];
            aindex_t p_v = ptxt[(*plen) - i - 1];
            // cout << c_v << ", ";
            hits[c_v] = hits[c_v] -  1;
            if (hits[c_v] == 0) {
                tbl[p_v].erase(c_v);
                (*seen).erase(c_v);
            }
        }
        (*plen) = (*plen) - num;
     
        // pp_tbl();
        // pp_seen();
    }
    int advance() {
        bool works = false;
        int found = 0;
        // int wdi = 0;
        
        int wdi = wstate[(*wslen)];
     
        
        while ((!works) & (wdi<WORD_DICT_SIZE)) {
            works = true;
            //if ((*wslen==56) & /* (wstate[2] == 13) &  (wstate[3] == 12) & */ (wdi==20)) {
            //    __debugbreak();
            //}
            for (auto ci = 0; ci < wlens[wdi]+1; ci++) {
                aindex_t ch = 0;
                if (ci < wlens[wdi]) {
                    ch = words[wdi][ci];
                }
                // cout << "Consuming " << id2c[ch] << " at " << (*plen) << endl;
                //  pp_tbl();
                bool cons = consume(ch);
                // cout << "cons: " << cons << "\n";
                // pp_tbl();
                if (!cons) {
                    works = false;
                    rollback(ci);
                    break;
                }
                else {
                   
                        if ((*plen) == TXT_LEN) {
                            found = true;
                            return 1;
                        }
                    
                }
            }
            if (works) {
                // print_ptxt(ptxt,*plen);
                break;
            }
            else {
            wdi++;
            }
        }
        if (works) {
            wstate[*wslen] = wdi;
            (*wslen)++;
            return 0;
        }
        else {
            
            int wwdi = wstate[(*wslen) - 1];
            wstate[(*wslen)] = 0;
            while (wwdi > (WORD_DICT_SIZE - 2)) {
                rollback(wlens[wstate[(*wslen) - 1]] + 1);
                wstate[(*wslen) - 1] = 0;
                (*wslen)--;
                if ((*wslen) < 1) {
                    return -1;
                }
                wwdi = wstate[(*wslen) - 1];
            }
            rollback(wlens[wstate[(*wslen) - 1]] + 1);
            wstate[(*wslen) - 1]++;
            (*wslen)--;
            return advance();
        }
        
    }
    
};



void decrypt_type2() {
    int wstate[TXT_LEN] = {};
    int wslen = 0;
    aindex_t ptxt[TXT_LEN] = {};
    int plen = 0;
    int hits[ROW_SUM] = {};
    set<aindex_t> tbl[27];
    map<aindex_t, aindex_t> seen{};
    State state(wstate,&wslen,ptxt,&plen,hits,tbl,&seen);
    int found = 0;
    while (found == 0) {
        found = state.advance();
    }
    print_ptxt(state.ptxt);
}

mutex p_mutex;

void find_th(int tnum, int idx) {
    int delta = WORD_DICT_SIZE / (double)tnum;

    int wstate[TXT_LEN] = {};
    int wslen = 0;
    wstate[0] = idx * delta;
    aindex_t ptxt[TXT_LEN] = {};
    int plen = 0;
    int hits[ROW_SUM] = {};
    set<aindex_t> tbl[27];
    map<aindex_t, aindex_t> seen{};
    State state(wstate, &wslen, ptxt, &plen, hits, tbl, &seen);
    int found = 0;
    while (found == 0) {
        found = state.advance();
    }
    if (found == 1) {
    p_mutex.lock();
    print_ptxt(state.ptxt);
    exit(0);
    }
    return;
}

/* thread number should be power of 2 */
void decrypt_type2_mt(int threads = 16) {
    if (threads == 1) {
        return decrypt_type2();
    }
    vector<thread> threadz{};
    for (auto i = 0; i < threads; i++) {
        threadz.push_back(thread(find_th,threads,i));
    }
    for (auto i = 0; i < threads; i++) {
        threadz[i].join();
    }
}

void decrypt_main(int t) {
    if (!decrypt_type1(cipher)) {
        // Inconclusive
        // cout << "NOT IN PDIC!" << endl;
    }
    decrypt_type2_mt(t);
}



int main(int argc, char* argv[])
{
    int t = 1;
    if (argc == 3) {
        // SHOULD be -t <thread_number>
        t = stoi(argv[2]);
        t = to_nearest_pw2(t);
    }
    else if (argc == 2) {
        // SHOULD be -mt
        t = thread::hardware_concurrency();
        t = to_nearest_pw2(t);
        if (t == 0) {
            t = 1;
        }
    }

    read_cipher(cipher);
    decrypt_main(t);
}

File: python/Decry.py
from os import stat


c2id = {' ':0,'a':1,'b':2,'c':3,'d':4,'e':5,'f':6,'g':7,'h':8,'i':9,'j':10,'k':11,'l':12,'m':13,'n':14,'o':15,'p':16,'q':17,'r':18,'s':19,'t':20,'u':21,'v':22,'w':23,'x':24,'y':25, 'z':26}
alphabet = " abcdefghijklmnopqrstuvwxyz"
T = 106 # Sum(rowlen)
rowlen = [19,7,1,2,4,10,2,2,5,6,1,1,3,2,6,6,2,1,5,5,7,2,1,2,1,2,1]
rowlen_d = dict()
for i,rl in enumerate(rowlen):
    rowlen_d[alphabet[i]] = rl

words = [x.strip() for x in open("./data/word.txt").readlines()]
ptxt_dict = [x for x in open("./data/plaintxts.txt").readlines()]

def checker(cipher,ptxt):
    # print(ptxt)
    k = dict()
    ktbl = dict()
    for a in alphabet:
        k[a] = set()
    for i,c in enumerate(cipher):
        pc = ptxt[i]
        k[pc].add(c)
        if c not in ktbl:
            ktbl[c] = pc
        else:
            c_map = ktbl[c]
            if c_map != pc:
                # print(f"NOT SAME! CAUGHT AT {i}")
                return -1
        if len(k[pc])>rowlen_d[pc]:
            # print(f"NOT SAME! CAUGHT AT {i}")
            return -1
    return 1

cipher = [int(x) for x in input().split(",")]
# print(cipher)

def gen_empty_key():
    k = dict()
    for i in range(27):
        k[i] = set()
    return k

class State:
    def __init__(self):
        self.ptxt = []
        for i in range(500):
            self.ptxt.append(0)
        self.wstate = []
        for i in range(500):
            self.wstate.append(0)
        self.wslen = 0
        self.plen = 0
        self.hits = []
        for i in range(T):
            self.hits.append(0)
        self.tbl = gen_empty_key()
        self.seen = dict()

    def consume(self,plc):
        cv = cipher[self.plen]
        if(self.hits[cv]>0):
            pe = self.seen[cv]
            if pe!= plc:
                return False
            else:
                self.hits[cv] += 1
                self.ptxt[self.plen] = pe
                self.plen += 1
                return True
        else:
            if len(self.tbl[plc])+1>rowlen[plc]:
                return False
            else:
                self.seen[cv] = plc
                self.tbl[plc].add(cv)
                self.hits[cv] += 1
                self.ptxt[self.plen] = plc
                self.plen += 1
                return True

    def rollback(self,num):
        # print(self.tbl)
        # print(self.seen)
        for i in range(num):
            cv = cipher[self.plen -i - 1]
            pv = self.ptxt[self.plen -i - 1]
            self.hits[cv] = self.hits[cv] - 1
            if self.hits[cv] == 0:
                self.tbl[pv].remove(cv)
                self.seen.pop(cv)
        self.plen -= num

    def advance(self):
        works = False
        found = 0
        wdi = self.wstate[self.wslen]
        
        # xxd = ""
        # for i in range(self.wslen):
        #     xxd += f"{words[self.wstate[i]]} "
        #     # print(f"{words[self.wstate[i]]} ",end="")
        # # print("\n")
        # if xxd.startswith("dislocating"):
        #     breakpoint()

        while (not works) and (wdi<len(words)):
            works = True
            for ci in range(len(words[wdi])+1):
                ch = 0
                if ci<len(words[wdi]):
                    ch = c2id[words[wdi][ci]]
                cons = self.consume(ch)
                if not cons:
                    works = False
                    self.rollback(ci)
                    break
                else:
                    if self.plen == 500:
                        found = 1
                        return 1
            
            if works:
                break
            else:
                wdi+=1
        
        if works:
            self.wstate[self.wslen] = wdi
            self.wslen += 1
            return 0

        else:
            wwdi = self.wstate[self.wslen-1]
            self.wstate[self.wslen] = 0
            while wwdi> (len(words)-2):
                self.rollback(len(words[self.wstate[self.wslen-1]])+1)
                self.wstate[self.wslen-1] = 0
                self.wslen -= 1
                if self.wslen < 1:
                    return -1
                wwdi = self.wstate[self.wslen-1]
            self.rollback(len(words[self.wstate[self.wslen-1]])+1)
            self.wstate[self.wslen - 1] += 1
            self.wslen -= 1
            return self.advance()


def decrypt():
    state = State()
    found = 0
    while found==0:
        found = state.advance()
    # print(found)
    for i,pc in enumerate(state.ptxt):
        print(alphabet[pc],end="")
    print("\n")

import time
t0 = time.time_ns()
for p in ptxt_dict:
    if checker(cipher,p)!=-1:
        print(p)
        exit(0)
decrypt()
print(time.time_ns()-t0)